crypto/fipsmodule/bn/montgomery.c

```diff
@@ -119,8 +119,37 @@
 #include <openssl/type_check.h>
 
 #include "internal.h"
+#include "../cpucap/internal.h"
 #include "../../internal.h"
 
+#if !defined(OPENSSL_NO_ASM) &&                                                \
+    (defined(OPENSSL_LINUX) || defined(OPENSSL_APPLE)) &&                      \
+    defined(OPENSSL_AARCH64) && defined(OPENSSL_BN_ASM_MONT)
+
+#include "../../../third_party/s2n-bignum/include/s2n-bignum_aws-lc.h"
+
+#define BN_MONTGOMERY_USE_S2N_BIGNUM 1
+
+OPENSSL_INLINE int montgomery_use_s2n_bignum(unsigned int num) {
+  // Use s2n-bignum's functions only if
+  // (1) The ARM architecture has slow multipliers, and
+  // (2) num (which is the number of words) is multiplie of 8, because
+  //     s2n-bignum's bignum_emontredc_8n requires it
+  // (3) The word size is 64 bits, and
+  // (4) Temporary buffer's size (t and mulres) used in
```

Comment:
what I still don't understand is if this is a limitation of s2n-bignum or aws-lc?

Reply:
I think the necessity of buffer comes from how the montgomery multiplication is being done in this performance improvement.
First we calculate the NxN->2N big int multiplication using Karatsuba multiplication, and then reduce 2N to N using montgomery reduction.
In order to do this, we need 2N-word temporary buffer.
Also, Karatsuba multiplication itself requires storing its intermediate result in a buffer, adding more usage.

Reply:
> s2n-bignum may use slightly more buffer
but why can't we just allocate large enough `t` and `mulres`? Are `s2n-bignum` functions that are used here writing to any other buffer?